Add Ecdsa adaptor and schnorr signature from temp-everything branch

Tibo-lg opened the pull request:
This PR adds binding for ecdsa adaptor signature and schnorr signature from [this](https://github.com/nkohen/secp256k1/tree/temp-everything) secp256k1 branch.

First commit is mainly auto-generated code (plus some changes to the script that do the autogeneration).
Second commit are changes to secp256k1-sys
Third commit is the higher level layer adding the types and functions.

Tibo-lg committed:
Update secp256k1-sys to add adaptor module

Tibo-lg committed:
Update rust-secp256k1 to add ecdsa-adaptor

elichai committed:
Add AligneType and redo secp256k1_context_create with alloc

elichai committed:
Replace use of boxes with global allocator

elichai committed:
Make preallocated use AlignedType

elichai committed:
Adopt no-std tests to new preacllocated_* functions

nkohen commented on src/schnorrsig.rs:
```diff
@@ -0,0 +1,435 @@
+        &self,
+        msg: &Message,
+        sk: &SecretKey,
+        nonce: &SchnorrNonce,
```
Actually I think this is meant to be a private key, `k`, whereas the type `SchnorrNonce` represents public keys `R = k*G`. The name is kind of confusing because both `k` and `R` values are sometimes called the nonce...

Tibo-lg replied:
Good catch thanks! I indeed got confused because it was called nonce, changed to private key.

Tibo-lg committed:
Update dependency folder to use ecdsa adaptor

Tibo-lg committed:
Add ecdsa adaptor example

vorot93 committed:
global-context depends on rand-std

justinmoon committed:
Implement lexigraphic ordering for PubKey

apoelstra committed:
switch from travis to github workflows

apoelstra committed:
Merge pull request #246 from vorot93/patch-2

global-context depends on rand-std

apoelstra committed:
Merge pull request #250 from rust-bitcoin/2020-11--github-actions

switch from travis to github workflows

justinmoon committed:
Fix link in README

apoelstra committed:
Merge pull request #252 from justinmoon/readme

Fix link in README

apoelstra committed:
Merge pull request #248 from justinmoon/pubkey-ordering

Implement lexigraphic ordering for PubKey

apoelstra committed:
make `ffi::Type::new` functions all unsafe, expand documentation

apoelstra committed:
Merge pull request #237 from p2pderivatives/add-bip340-schnorr

Add bip340 schnorr

apoelstra committed:
schnorrsig: expose tweak_check_add API

apoelstra committed:
stop explicitly casting references to rawptrs

apoelstra committed:
schnorrsig: change tweak_add_check to return a bool, take a fixed-length array

apoelstra committed:
Merge pull request #253 from apoelstra/2020-11--schnorrsig-followup

BIP 0340 followups

elichai committed:
Make test.sh test and build secp256k1-sys

Tibo-lg committed:
Implement low r signing

apoelstra committed:
Merge pull request #233 from elichai/alloc-AlignedType2

Making sure everything is aligned correctly. Succeeder of #141

apoelstra committed:
bump secp-sys version; drop endomorphism flag

apoelstra committed:
secp-sys: update scratch_impl.patch for upstream changes

apoelstra committed:
rust-secp 0.20.0

thomaseizinger committed:
Run more tests in wasm

Instead of repeating ourselves in defining one big test for the wasm
target, we can override the `test` attribute with the `wasm-bindgen-test`
one and therefore automatically run all (supported) tests in wasm.

Unfortunately, wasm doesn't support catching panics yet which means we
have to disable the `test_panic_raw_ctx` test.

tcharding committed:
Allow unnecessary parentheses

Clippy emits:

	warning: unnecessary parentheses around assigned value

Add a attribute to allow unnecessary parentheses.

tcharding committed:
Conditionally include ALIGN_TO

This const is only used under specific features, use `cfg` to
conditionally build it in.

Removes clippy warning.

tcharding committed:
Conditionally compile fn strlen

`strlen` is only used under certain feature flags, use `cfg` to
conditionally build it in.

Clears clippy warning.

tcharding committed:
Implement AsRef instead of custom method

Clippy emits a warning since we define a method that has the same name
as a standard trait. Implement the trait `AsRef` instead of using a
custom method.

apoelstra committed:
Merge pull request #260 from tcharding/clippy

Clear clippy warnings

tcharding committed:
Implement is_empty method

Clippy warns of missing `is_empty`, trivially implement it by calling
through to `self.data.is_empty()`.

tcharding committed:
Remove unnecessary return statements

Found by clippy. We don't need a `return` for the final statement.

tcharding committed:
Remove unused error return value

This helper never returns an error, remove the `Result` return type.
Found by clippy.

tcharding committed:
Return Ok directly

Clippy emits warning:

	warning: passing a unit value to a function

Just return `Ok(())` after calling `fill_bytes`.

tcharding committed:
Use for loop instead of map

Currently we are misusing `map` on an iterator to loop `n` times,
additionally the assertion is pointless. Use a for loop and assert
against the length of the set.

tcharding committed:
Remove unnecessary clone

Type is `Copy`, no need for clone.

tcharding committed:
Use ManuallyDrop

Suggested by clippy, we need to use ManuallyDrop for these types in
order to correctly free up the memory.

apoelstra committed:
Merge pull request #261 from tcharding/more-clippy

More clippy

apoelstra committed:
Merge pull request #258 from thomaseizinger/better-wasm-tests

Run more tests in wasm

apoelstra committed:
replace cargo `external-symbols` feature with a rustc --cfg flag

This feature was not useful for Cargo users, since Cargo does not give you
the kind of fine-grained control over C library linkage that you need. So
it was just unnecessarily confusing and would cause the build to break if
you enabled it accidentally, say, with --all-features.

apoelstra committed:
replace `fuzztarget` Cargo feature with a rustc --cfg flag

It's super dangerous to use Cargo features for this, since they can be set
accidentally (or maliciously by any crate in a user's entire dep tree). Instead
we can just require users set `RUSTFLAGS` appropriately, which we can easily
do in our fuzzing scripts.

apoelstra committed:
add warning about fuzzing being enabled

apoelstra committed:
fuzz: only replace signing and verification, leave everything else alone

We can now run unit tests with the fuzz feature on, and they'll pass,
which is some assurance that fuzzing with the feature on won't lead to
spurious failures due to the fuzz harness inadequately simulating message
signing.

apoelstra committed:
ffi: make function types nullable

apoelstra committed:
Merge pull request #263 from apoelstra/2020-12--no-extsymb

Replace dangerous cargo features with rustc flags

apoelstra committed:
Merge pull request #259 from p2pderivatives/implement-low-r-signing

Implement low r signing

apoelstra committed:
Merge pull request #256 from rust-bitcoin/workspace

Make test.sh test and build secp256k1-sys

apoelstra committed:
fuzz: implement recoverable signatures, get all tests passing, run them in CI

junderw committed:
Fix link_name for secp256k1_ec_seckey_negate

apoelstra committed:
fuzz: disable sig-grinding tests, prevent spin-looping

apoelstra committed:
fuzz: fix elichai's nits

apoelstra committed:
Merge pull request #264 from rust-bitcoin/2020-12--no-extsymb2

redo fuzz target

apoelstra committed:
secp-sys:  remove now-unneeded endianness check from build.rs

apoelstra committed:
Merge pull request #266 from junderw/patch-1

Fix link_name for secp256k1_ec_seckey_negate

apoelstra committed:
update libsecp to 98dac87839838b86094f1bccc71cc20e67b146cc

apoelstra committed:
drop unnecessary no_mangle annotations on non-exported functions

TheBlueMatt committed:
Add ability to randomize context without the `rand` feature.

There is little reason to pull in the `rand` dep just for the `Rng`
trait for users who want to randomize contexts. We should expose a
randomize function that just takes 32 bytes.

apoelstra committed:
Merge pull request #267 from TheBlueMatt/master

Add ability to randomize context without the `rand` feature.

apoelstra committed:
Merge pull request #257 from apoelstra/2020-12--no-endo

update libsecp, remove endomorphism feature flag and release 0.20.0

apoelstra committed:
rename `rust_secp_fuzz` to `fuzzing`

apoelstra committed:
bump version to 0.20.1

apoelstra committed:
Merge pull request #269 from apoelstra/2021-01--rename-fuzz-feature

Rename `rust_secp_fuzz` config flag to `fuzzing`

gregdhill committed:
use core instead of std for wasm sanity checks

Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

apoelstra committed:
Merge pull request #275 from gregdhill/master

use core instead of std for wasm sanity checks

dr-orlovsky committed:
More standard derives for Error

apoelstra committed:
Merge pull request #277 from LNP-BP/fix/error-derive

More standard derives for Error

fanatid committed:
Use CARGO_CFG_TARGET_ARCH instead TARGET in build.rs

apoelstra committed:
Merge pull request #284 from fanatid/patch-1

Use CARGO_CFG_TARGET_ARCH instead TARGET in build.rs

Tibo-lg committed:
Add combine_keys function to PublicKey
